x/gamm/twap/store_test.go

```diff
@@ -0,0 +1,181 @@
+	defaultRevInputAt := func(t time.Time) getRecordInput { return getRecordInput{1, t, "tokenA", "tokenB"} }
+	baseRecord := newEmptyPriceRecord(1, baseTime, "tokenB", "tokenA")
+	tMin1 := baseTime.Add(-time.Second)
+	tMin1Record := newEmptyPriceRecord(1, tMin1, "tokenB", "tokenA")
+	tPlus1 := baseTime.Add(time.Second)
+	tPlus1Record := newEmptyPriceRecord(1, tPlus1, "tokenB", "tokenA")
+
+	tests := map[string]struct {
+		recordsToSet   []types.TwapRecord
+		input          getRecordInput
+		expectedRecord types.TwapRecord
+		expErr         bool
+	}{
+		"no entries":            {[]types.TwapRecord{}, defaultInputAt(baseTime), baseRecord, true},
+		"get at latest (exact)": {[]types.TwapRecord{baseRecord}, defaultInputAt(baseTime), baseRecord, false},
+		"rev at latest (exact)": {[]types.TwapRecord{baseRecord}, defaultRevInputAt(baseTime), baseRecord, true},
+
+		"get latest (exact) w/ past entries": {
+			[]types.TwapRecord{tMin1Record, baseRecord}, defaultInputAt(baseTime), baseRecord, false},
+		"get entry (exact) w/ a subsequent entry": {
+			[]types.TwapRecord{tMin1Record, baseRecord}, defaultInputAt(tMin1), tMin1Record, false},
+		"get sandwitched entry (exact)": {
+			[]types.TwapRecord{tMin1Record, baseRecord, tPlus1Record}, defaultInputAt(baseTime), baseRecord, false},
+		"rev sandwitched entry (exact)": {
+			[]types.TwapRecord{tMin1Record, baseRecord, tPlus1Record}, defaultRevInputAt(baseTime), baseRecord, true},
+
+		"get future":                 {[]types.TwapRecord{baseRecord}, defaultInputAt(tPlus1), baseRecord, false},
+		"get future w/ past entries": {[]types.TwapRecord{tMin1Record, baseRecord}, defaultInputAt(tPlus1), baseRecord, false},
+
+		"get in between entries (2 entry)": {
+			[]types.TwapRecord{tMin1Record, baseRecord},
+			defaultInputAt(baseTime.Add(-time.Millisecond)), tMin1Record, false},
+		"get in between entries (3 entry)": {
+			[]types.TwapRecord{tMin1Record, baseRecord, tPlus1Record},
+			defaultInputAt(baseTime.Add(-time.Millisecond)), tMin1Record, false},
+		"get in between entries (3 entry) #2": {
+			[]types.TwapRecord{tMin1Record, baseRecord, tPlus1Record},
+			defaultInputAt(baseTime.Add(time.Millisecond)), baseRecord, false},
+
+		"query too old": {
+			[]types.TwapRecord{tMin1Record, baseRecord, tPlus1Record},
+			defaultInputAt(baseTime.Add(-time.Second * 2)),
+			baseRecord, true},
+
+		"non-existent pool ID": {
+			[]types.TwapRecord{tMin1Record, baseRecord, tPlus1Record},
+			wrongPoolIdInputAt(baseTime), baseRecord, true},
+	}
+	for name, test := range tests {
+		s.Run(name, func() {
+			s.SetupTest()
+			for _, record := range test.recordsToSet {
+				s.twapkeeper.StoreNewRecord(s.Ctx, record)
+			}
+			record, err := s.twapkeeper.GetRecordAtOrBeforeTime(
+				s.Ctx,
+				test.input.poolId, test.input.t, test.input.asset0Denom, test.input.asset1Denom)
+			if test.expErr {
+				s.Require().Error(err)
```

Comment:
consider asserting on what error specifically is returned

Reply:
I don't think thats important personally